src/codecs/qoi.rs

```diff
@@ -0,0 +1,97 @@
+//! Decoding and encoding of QOI images
+
+use crate::{
+    error::{DecodingError, EncodingError},
+    ColorType, ImageDecoder, ImageEncoder, ImageError, ImageFormat, ImageResult,
+};
+use std::io::{Cursor, Read, Write};
+
+/// QOI decoder
+pub struct QoiDecoder<R> {
+    decoder: qoi::Decoder<R>,
+}
+
+impl<R> QoiDecoder<R>
+where
+    R: Read,
+{
+    /// Creates a new decoder that decodes from the stream ```reader```
+    pub fn new(reader: R) -> ImageResult<Self> {
+        let decoder = qoi::Decoder::from_stream(reader).map_err(decoding_error)?;
+        Ok(Self { decoder })
+    }
+}
+
+impl<'a, R: Read + 'a> ImageDecoder<'a> for QoiDecoder<R> {
+    type Reader = Cursor<Vec<u8>>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        (self.decoder.header().width, self.decoder.header().height)
+    }
+
+    fn color_type(&self) -> ColorType {
+        match self.decoder.header().channels {
+            qoi::Channels::Rgb => ColorType::Rgb8,
+            qoi::Channels::Rgba => ColorType::Rgba8,
+        }
+    }
+
+    fn into_reader(mut self) -> ImageResult<Self::Reader> {
+        let buffer = self.decoder.decode_to_vec().map_err(decoding_error)?;
+        Ok(Cursor::new(buffer))
+    }
+}
+
+fn decoding_error(error: qoi::Error) -> ImageError {
+    ImageError::Decoding(DecodingError::new(ImageFormat::Qoi.into(), error))
+}
+
+fn encoding_error(error: qoi::Error) -> ImageError {
+    ImageError::Encoding(EncodingError::new(ImageFormat::Qoi.into(), error))
+}
+
+/// QOI encoder
+pub struct QoiEncoder<W> {
+    writer: W,
+}
+
+impl<W: Write> QoiEncoder<W> {
+    /// Creates a new encoder that writes its output to ```writer```
+    pub fn new(writer: W) -> Self {
+        Self { writer }
+    }
+}
+
+impl<W: Write> ImageEncoder for QoiEncoder<W> {
+    fn write_image(
+        mut self,
+        buf: &[u8],
+        width: u32,
+        height: u32,
+        _color_type: ColorType,
```

Comment:
This one is still open.

> Ignoring color isn't correct here. Rather, QOI supports only 8-bit rgb-ish color (with optional colorspace). We'll want to either convert here, or raise on error on unsupported formats.


https://github.com/image-rs/image/pull/1780#discussion_r960338581